### PR TITLE
[CARBONDATA-2587][CARBONDATA-2588] Local Dictionary Data Loading support

### PR DESCRIPTION
**Pull request by kumarvishal09**
**What changes are proposed in this PR**

1. Added code to support Local Dictionary Data Loading for primitive type
2. Added code to support Local Dictionary Data Loading for complex type.

**How this PR is tested**
Manual testing is done in 3 Node setup.
UT will be raised in different PR

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5314/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6483/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6484/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5315/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6485/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5316/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5405/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5406/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5407/

File: core/src/main/java/org/apache/carbondata/core/datastore/page/LocalDictColumnPage.java [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+  public PageLevelDictionary getPageDictionary() {
+    return pageLevelDictionary;
+  }
+
+  @Override public void disableLocalDictEncoding() {
+    pageLevelDictionary = null;
+    freeEncodedColumnPage();
+  }
+
+  @Override public PageLevelDictionary getColumnPageDictionary() {
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/datastore/page/LocalDictColumnPage.java]: duplicate method are present, remove one `getColumnPageDictionary` and `getPageDictionary`

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java [hunk 2]
```diff
@@ -482,7 +482,7 @@ public String getTableUniqueName() {
    * @return
    */
   public boolean isLocalDictionaryEnabled() {
-    return isLocalDictionaryEnabled;
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: why it is always false

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/util/ExampleUtils.scala [hunk 3]
```diff
@@ -96,8 +96,8 @@ object ExampleUtils {
     import spark.implicits._
     val sc = spark.sparkContext
     val df = sc.parallelize(1 to numRows, 2)
-        .map(x => ("a", "b", x))
```
Comment (akashrn5) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/util/ExampleUtils.scala]: remove the unnecessary changes

File: processing/src/main/java/org/apache/carbondata/processing/store/TablePage.java [hunk 4]
```diff
@@ -104,19 +105,27 @@
         page.setStatsCollector(KeyPageStatsCollector.newInstance(DataTypes.BYTE_ARRAY));
         dictDimensionPages[tmpNumDictDimIdx++] = page;
       } else {
+        // will be encoded using string page
+        LocalDictionaryGenerator localDictionaryGenerator =
+            model.getColumnLocalDictGenMap().get(spec.getFieldName());
         if (DataTypes.VARCHAR == spec.getSchemaDataType()) {
-          page = ColumnPage.newPage(spec, DataTypes.VARCHAR, pageSize);
+          page = ColumnPage.newLocalDictPage(spec,
```
Comment (akashrn5) [inline processing/src/main/java/org/apache/carbondata/processing/store/TablePage.java]: if the `localDictionaryGenerator` is null, for VARCHAR also, it should not generate localDictPage, null check is missing i think

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java (hunk 2 above)
Comment (kumarvishal09) [reply core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: Currently Query part is not handled, so query will fail if local dictionary is enabled. I am working on the same

File: core/src/main/java/org/apache/carbondata/core/datastore/page/LocalDictColumnPage.java (hunk 1 above)
Comment (kumarvishal09) [reply core/src/main/java/org/apache/carbondata/core/datastore/page/LocalDictColumnPage.java]: remove above method

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5414/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5325/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6494/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5415/

**Comment by kumarvishal09** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5416/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6497/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5328/

File: core/src/main/java/org/apache/carbondata/core/datastore/blocklet/BlockletEncodedColumnPage.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+   */
+  private ArrayDeque<Future<FallbackEncodedColumnPage>> fallbackFutureQueue;
+
+  BlockletEncodedColumnPage(ExecutorService fallbackExecutorService,
+      EncodedColumnPage encodedColumnPage) {
+    this.encodedColumnPageList = new ArrayList<>();
+    this.fallbackExecutorService = fallbackExecutorService;
+    this.encodedColumnPageList.add(encodedColumnPage);
+    // if dimension page is local dictionary enabled and encoded with local dictionary
+    if (encodedColumnPage.isLocalDictionaryEnabled() && encodedColumnPage
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/blocklet/BlockletEncodedColumnPage.java]: Just keep `this.isLocalDictEncoded =encodedColumnPage.isLocalDictGeneratedPage()` should be ok

File: core/src/main/java/org/apache/carbondata/core/datastore/blocklet/BlockletEncodedColumnPage.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+   */
+  private PageLevelDictionary pageLevelDictionary;
+
+  /**
+   * fallback future task queue;
+   */
+  private ArrayDeque<Future<FallbackEncodedColumnPage>> fallbackFutureQueue;
+
+  BlockletEncodedColumnPage(ExecutorService fallbackExecutorService,
+      EncodedColumnPage encodedColumnPage) {
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/blocklet/BlockletEncodedColumnPage.java]: Don't add `encodedColumnPage` from constructor, use `addEncodedColumnColumnPage`

File: core/src/main/java/org/apache/carbondata/core/datastore/blocklet/BlockletEncodedColumnPage.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+   * @return list of encoded page list
+   */
+  public List<EncodedColumnPage> getEncodedColumnPageList() {
+    // if fallback queue is null then for some pages fallback was initiated
+    if (null != this.fallbackFutureQueue) {
+      try {
+        // check if queue is not empty
+        while (!fallbackFutureQueue.isEmpty()) {
+          // get the head element of queue
+          Future<FallbackEncodedColumnPage> fallbackTask = fallbackFutureQueue.getFirst();
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/datastore/blocklet/BlockletEncodedColumnPage.java]: I think just use `poll` and `future.get` to fill the pages instead of using `sleep`

File: core/src/test/java/org/apache/carbondata/core/util/CarbonMetadataUtilTest.java [hunk 8]
```diff
@@ -172,71 +172,71 @@
...
-    BlockletMinMaxIndex blockletMinMaxIndex = new BlockletMinMaxIndex();
-    blockletMinMaxIndex.addToMax_values(ByteBuffer.wrap(byteMaxArr));
-    blockletMinMaxIndex.addToMin_values(ByteBuffer.wrap(byteMinArr));
-    FileFooter3 footer = convertFileFooterVersion3(blockletInfoColumnarList,
-        indexList,
-        cardinality, 2);
-    assertEquals(footer.getBlocklet_index_list(), indexList);
-
-  }
+//
```
Comment (ravipesala) [inline core/src/test/java/org/apache/carbondata/core/util/CarbonMetadataUtilTest.java]: remove if not required

File: processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java [hunk 9]
```diff
@@ -623,5 +640,83 @@ public DataMapWriterListener getDataMapWriterlistener() {
...
+  }
+
+  /**
+   * This method prepares a map which will have column and local dictionary generator mapping for
+   * all the local dictionary columns.
+   * @param carbonTable
+   * @param wrapperColumnSchema
+   * @param carbonFactDataHandlerModel
+   */
+  public static void setLocalDictToModel(CarbonTable carbonTable,
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java]: Keep as `private`

File: core/src/main/java/org/apache/carbondata/core/localdictionary/dictionaryholder/MapBasedDictionaryStore.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+   * to maintain dictionary key value
+   */
+  private final Map<DictionaryByteArrayWrapper, Integer> dictionary;
+
+  /**
+   * maintaining array for reverse lookup
+   * otherwise iterating everytime in map for reverse lookup will be slowdown the performance
+   * It will only maintain the reference
+   */
+  private byte[][] referenceDictionaryArray;
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/localdictionary/dictionaryholder/MapBasedDictionaryStore.java]: Better directly use `DictionaryByteArrayWrapper` array here

File: core/src/main/java/org/apache/carbondata/core/localdictionary/PageLevelDictionary.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    dictionaryColumnPage.setStatsCollector(new DummyStatsCollector());
+    int rowId = 0;
+    //starting index is 1 as dictionary value starts from 1
+    for (int i = 1; i <= maxDictValue; i++) {
+      if (usedDictionaryValues.get(i)) {
+        dictionaryColumnPage
+            .putData(rowId++, localDictionaryGenerator.getDictionaryKeyBasedOnValue(i));
+      } else {
+        dictionaryColumnPage
+            .putData(rowId++, CarbonCommonConstants.EMPTY_BYTE_ARRAY);
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/localdictionary/PageLevelDictionary.java]: Check any other cases in data we get empty binary.

File: processing/src/main/java/org/apache/carbondata/processing/store/writer/v3/CarbonFactDataWriterImplV3.java [hunk 12]
```diff
@@ -76,7 +79,29 @@ public CarbonFactDataWriterImplV3(CarbonFactDataHandlerModel model) {
       blockletSizeThreshold = fileSizeInBytes;
       LOGGER.info("Blocklet size configure for table is: " + blockletSizeThreshold);
     }
-    blockletDataHolder = new BlockletDataHolder();
+    int numberOfCores = CarbonProperties.getInstance().getNumberOfCores();
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/store/writer/v3/CarbonFactDataWriterImplV3.java]: please remove unused code

File: processing/src/main/java/org/apache/carbondata/processing/store/writer/v3/CarbonFactDataWriterImplV3.java [hunk 13]
```diff
@@ -110,44 +135,51 @@ public CarbonFactDataWriterImplV3(CarbonFactDataHandlerModel model) {
...
-      boolean isAdded = false;
-      // check if size more than blocklet size then write the page to file
-      if (blockletDataHolder.getSize() + tablePage.getEncodedTablePage().getEncodedSize() >=
-          blockletSizeThreshold) {
-        // if blocklet size exceeds threshold, write blocklet data
-        if (blockletDataHolder.getEncodedTablePages().size() == 0) {
-          isAdded = true;
-          addPageData(tablePage);
-        }
+    try {
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/store/writer/v3/CarbonFactDataWriterImplV3.java]: dnt format the code if code is not changed

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6536/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5365/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5440/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5441/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5366/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6557/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5386/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6566/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5464/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6569/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5397/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5474/
